### PR TITLE
Update to context-mod->props

### PR DESCRIPTION
**Pull request by stroncaro**
This pull request contains the changes that Urs and I discussed in the Lilypond mailing list.

Summary of changes:
1) The ruleset is a list that begins with either 'strict or 'flexible, which determines if arguments not specified in the ruleset are rejected (strict mode) or accepted (flexible mode). The enforcement symbol is followed by any number of rules.
2) A rule can have any of the following forms:
   - arg-name: a symbol, stating that this argument is required.
   - (arg-name): a list with a symbol, stating that this argument is required.
   - (arg-name ,type?): same as above with procedure for a type checking.
   - (? arg-name): a list with the symbol '? and another symbol, stating that the argument is optional.
   - (? arg-name ,type?): same as above with a type-check.
   - (? arg-name ,type? def-v): same as above with a default value.
3) context-mod->props now only requires two arguments instead of three: an optional ruleset and a context-mod. The previously optional 'strict? boolean is now handled by the ruleset.
4) A macro for a convenient way of defining functions that use context-mod->props to handle options. It is called with "with-options" before the function definition, and with the ruleset before the body of the function.
Example:
```
\version "2.19.80"
\include "oll-core/package.ily"

testRules =
#(with-options define-void-function () () `(strict
      payload
      (something)
      (something-else)
      (target ,symbol?)
      (id ,number?)
      (? accepted-without-type)
      (? accepted-arg ,fraction?)
      (? ind ,number? 5)
      (? you-are-responsible-for-your-defaults ,number? "34")
      (? msg ,string? "No message given"))
   (pretty-print rules)
   (pretty-print props))

\testRules \with {
  msg = "Something"
  unk = "Unknown option"
  id = 3
  target = something
  accepted-arg = 7/4
  accepted-without-type = #(ly:make-moment 3/16)
  something-else = "something useful"
}
```
Also, the macro allows three ways of providing an empty ruleset:
```
emptyA =
#(with-options define-void-function () () ()
   (pretty-print rules))

emptyB =
#(with-options define-void-function () () '()
   (pretty-print rules))
emptyC =
#(with-options define-void-function () () `()
   (pretty-print rules))

\emptyA \with {
  function = "testing empty rulesets"
}

\emptyB \with {
  function = "testing empty rulesets"
}

\emptyC \with {
  function = "testing empty rulesets"
}
```

### CODE REVIEW

File: internal/options.ily [hunk 1]
```diff
@@ -105,45 +105,66 @@
...
+                  (f (first obj)))
+              (and (symbol? f)
+                   (or (= l 1)
+                       (and (= l 2)
+                            (if (eq? '? f)
+                                (symbol? (second obj))
+                                (procedure? (second obj))))
+                       (and (> l 2)
+                            (< l 5)
+                            (symbol? (second obj))
```
Comment (uliska) [inline internal/options.ily]: Shouldn't this also have the `(if (eq? '? f)` conditional like in the previous check?

File: internal/options.ily [hunk 2]
```diff
@@ -105,45 +105,66 @@
...
+                            (procedure? (third obj)))))))))
+
+#(define (enforcement-symbol? obj)
+  (or (eq? 'strict obj)
+      (eq? 'flexible obj)))
 
 #(define (prop-rules? obj)
    "Check if given object is a property rules structure.
-    This is true when obj is a list of 'prop-rule? entries"
+    This is true when obj is a list, its first element is an 'enforcement-symbol? and subsequent elements are 'prop-rule? entries"
```
Comment (uliska) [inline internal/options.ily]: Please add line break(s) somewhere

File: internal/options.ily [hunk 3]
```diff
@@ -105,45 +105,66 @@
...
     - type check
-    - Handle missing properties. If a default is avaible use that."
+    - Handle missing properties. If a default is available use that."
    (let*
-    ((rules
+    ((strict (eq? (car rules) 'strict))
+     (rules (cdr rules))
+     (rules
       (map (lambda (rule)
+             (if (symbol? rule) (set! rule (list rule)))
```
Comment (uliska) [inline internal/options.ily]: The usual style in oll would be to move this in the following `let` block as 

```scheme
(rule (if (symbol? rule) (list rule) rule))
```

File: internal/options.ily (hunk 1 above)
Comment (stroncaro) [reply internal/options.ily]: Now that you mention it, yes, and some kind of check for `'?` should also appear for the `(= l 1)` and the `(symbol? obj)` cases.

**Comment by stroncaro** [comment]
The changes are done and I also reworked `prop-rule?` which should be way more clear now (it was kind of convoluted for no purpose).

One thing that I was thinking just now that I had not considered is that of docstrings on the functions created with `with-options`. As it is now the first expression of the function is a `let` block, so I'm sure it's not working. But I don't know how to test that, do you know how? If so I could attempt to fix this issue.

**Comment by stroncaro** [comment]
I was thinking about the macro a bit more. Wouldn't it be useful if the macro defined all the arguments of props with their values so that they are usable out of the box? As it is now, it is still needed to create a `let` block to assign them. So what I mean is allow this:

```
someFuncWithOpts =
#(with-options define-void-function () ()
   `(strict
      (name ,string?)
      (? msg ,string? "No message given"))
   (pretty-print (format "~a says: ~a" name msg)))
```

**Comment by stroncaro** [comment]
I did not realize that module-handling.ily was using validate-props. This last commit takes care of the change of syntax.

**Comment by uliska** [comment]
Hi Stéfano, sorry for waiting quite some time with the review, but I didn't have the opportunity earlier.
I added a few commits - please check if I missed or misunderstood anything .

> One thing that I was thinking just now that I had not considered is that of docstrings on the functions created with with-options. As it is now the first expression of the function is a let block, so I'm sure it's not working. But I don't know how to test that, do you know how? If so I could attempt to fix this issue.

I'm not really sure what you mean with that. IIUC you can simply insert a (doc)string before the `let`. As it is now you have *one* expression in the macro (the `let`) but can't you just have *two* expressions with the first one being a literal (string), just like with any other procedure definition?
I have no clue about `define-macro`, but when I inserted a string nothing strange happened.

> Wouldn't it be useful if the macro defined all the arguments of props with their values so that they are usable out of the box? As it is now, it is still needed to create a let block to assign them. So what I mean is allow this:

This looks good, but as said above I have no idea about defining macros in Scheme. So if you have a solution please go ahead.

**Comment by stroncaro** [comment]
Hi Urs, the changes look good. I added only a slight modification because `prop-rule?` would reject optional arguments with a default value since it only considered a length of 2.


> I'm not really sure what you mean with that. IIUC you can simply insert a (doc)string before the let. As it is now you have one expression in the macro (the let) but can't you just have two expressions with the first one being a literal (string), just like with any other procedure definition?
> I have no clue about define-macro, but when I inserted a string nothing strange happened.

What I was thinking is that docstrings, AFAIK, are used for automatic generation of documentation or other things like that, but I'm not 100% sure because I've never used that functionality. If I understand it correctly, if the first expression of a procedure is a string, it is assigned to some kind of "meta" property of that procedure that can be accessed with some code.
However, as the macro is now, you can write a string as the first expression, and it will not generate any errors because a string is by itself a valid expression (that does nothing). But the macro will put this string as the first expression of it's `let*` block, after `rules` and `props` have already been defined, so the docstring will not get assigned as it should. I think I know how to make it work, but I don't know how to test this because I don't know how to retrieve the docstring from the proc. I'll look it up and make a fix later if it's needed.

> This looks good, but as said above I have no idea about defining macros in Scheme. So if you have a solution please go ahead.

I'll try it out.

Also, I was thinking of adding this helper function. I show it in an example:
```
\version "2.19.80"
\include "oll-core/package.ily"

#(define (props->arg-list-with-key props k)
   "Return a list with all the items in a `props obj that have the key `k."
   (map cdr
     (filter
       (lambda (x)
         (eq? (car x) k))
       props)))

overlap-opts =
#(with-options define-void-function () ()
   `(strict
      (chained ,string?))
   (let ((chain (props->arg-list-with-key props 'chained)))
     (pretty-print props)
     (pretty-print chain)))

\overlap-opts \with {
  chained = "All the vars"
  chained = "with the same name"
  chained = "are listed"
  chained = "one after the other."
}
```

**Pull request merged**